accepted/00000-maintenance-windows.md

```diff
@@ -0,0 +1,390 @@
+
+
+## Second Iteration
+
+### Implement a Maintenance Mode
+
+Put a system manually into "Maintenance Mode". This is useful in emergency situations where a system is down
+and needs to be fixed quickly.
+An API call should be implemented to do it. The required permissions needs to be discussed.
+Either a normal admin of the system could get the permissions to do this, or it still required organization admin permissions.
+This requires further discussion.
+
+Some companies might require to follow a workflow to put a machine into maintenance mode.
+To make this feature also usable for them, it should be possible to disable manual setting of the maintenance mode.
+
+Bringing a server back to production mode earlier than the maintenance window close is another feature which
+could be implemented in the second iteration.
+
+All this would require a trigger mechanism which execute changes on the system to reflect the maintenance mode.
+
+
+#### Emergency Maintenance Mode
+
+When a system is put into Maintenance Mode manually, it mostly happen because of an emergency case.
+It should allow to perform the necessary steps to fix the issue and bring the system or application
+back online.
+
+When a system is currently in Maintenance Mode, it should be possible to schedule actions "now" and for the next minute.
+
+
+#### Enforcing the Maintenance Mode on the client side
+
+With the proposal of the first iteration, we do not enforce the maintenance mode on system level.
+As long as nobody uses the salt bus directly, this is a non-issue.
+
+There needs to be discussions around how to implement such an enforcement.
+Using the current existing blackout mode for salt minions is too broad to differentiate between
+allowed and disallowed actions.
+
+The following actions needs be be considered:
+
+- Image Building (Container and Kiwi Images)
+- Create of new machines and Virtual Machines
+- Salt Boot
+
+All use `state.apply` but might be wanted to run outside of Maintenance Mode.
+
+As a solution we could enhance the minion blackout mode to check for metadata provided with the call.
+This would prevent "accidents", but would not really enforce the maintenance mode as everybody
+who has direct access to salt could set the required headers execute things when the system is not
+in Maintenance Mode.
+
+
+### Manage Maintenance Schedules
+
+It should be possible to add a URL to an iCalendar file and maybe implement CalDav support.
+
+
+### Assign a Maintenance Schedule to a System
+
+- adding a system to a Maintenance Schedule via activation key at onboarding time. This requires changes on the activation key.
+
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+
+# Alternatives
+[alternatives]: #alternatives
+
+# Unresolved questions
+[unresolved]: #unresolved-questions
+
+- performance: what happens when a lot of actions are scheduled at the beginning of a maintenance window? Will this create a performance problem? Is salt batching ready to handle such situations properly?
```

Comment:
We should be fine as long as we create one Action per type and timestamp with multiple clients associated. Specifically, we should try to avoid as much as possible the situation in which there is one Action per client.

Batching will limit the number of clients executing an Action (only considering that one Action).

My impression is that this proposal would not make the situation worse than it is now - of course users will still have to pay a bit of attention not to create too many separate Actions or batching won't help them. But by and large, they should be able to create Actions targeting many clients just like before - just "taking one pre-defined start datetime from a list" instead of a specifying one "freely".

We will have to pay attention when, for example, rescheduling Actions due to updates to their Maintenance Schedule - clients will have to stay grouped together as much as possible.